symengine/basic_to_upoly.cpp

```diff
@@ -0,0 +1,363 @@
+                            int i = positive_integer(powx->get_exp());
+                            if (i < 0)
+                                throw std::runtime_error(
+                                    "Could not extract generator");
+                            else if (i == 0)
+                                min_gen = pow(
+                                    min_gen,
+                                    div(one, rcp_static_cast<const Rational>(
+                                                 powx->get_exp())
+                                                 ->get_den()));
+                            // else was a positive integer, nothing to be done
+                            continue;
+                        }
+                    } else if (is_a<const Pow>(*min_gen)) {
+                        // won't handle cases like 4**x + 2**x
+                        RCP<const Pow> powx2
+                            = rcp_static_cast<const Pow>(min_gen);
+                        if (eq(*(powx->get_base()), *(powx2->get_base()))) {
+                            // won't handle things like 2**(3x/2) + 2**x, but
+                            // will handle 2**(x/2) + 2**x
+                            // TODO : can be modified to handle, will 
+                            // need some work
+                            // expand should not be needed here
+                            RCP<const Basic> tmp = expand(
+                                div(powx->get_exp(), powx2->get_exp()));
+                            int i = positive_integer(tmp);
+                            if (i > 0)
+                                continue;
+
+                            tmp = expand(
+                                div(powx2->get_exp(), powx->get_exp()));
+                            i = positive_integer(tmp);
+                            if (i > 0) {
+                                min_gen = curr_gen;
+                                continue;
+                            }
+                        }
+                    }
+
+                    throw std::runtime_error("Could not extract generator");
+                } else {
+                    throw std::runtime_error(
+                        "Internal Error : Wrong generator type");
+                }
+
+            } else {
+                SYMENGINE_ASSERT(is_a<const Symbol>(*curr_gen)
+                                 or is_a<const Pow>(*curr_gen)
+                                 or is_a_sub<const Function>(*curr_gen))
+                min_gen = curr_gen;
+                min_gen_set = true;
+            }
+        }
+
+        res = min_gen;
+    }
+
+    void bvisit(const Mul &x)
+    {
+        RCP<const Basic> gen;
+
+        if (not inpow)
+            if (not is_a<const Integer>(*x.coef_))
+                throw std::runtime_error("Non-integer coeff found");
+
+        if (x.dict_.size() > 1)
+            // this case should not be encountered, if `expand` was called
```

Comment:
This should be independent of `expand`. For example handling `(x+1)**10*(x+2)**10` using polynomials is faster than using `expand`